add new diagram; massive rewrite (#451)
mongodb/docs-app-services: source/sync/data-model/update-schema.txt

   * - :ref:`Remove an object type <schema-update-remove-object-type>`
     - Breaking
     - Non-breaking

>    * - :ref:`Remove an optional or required required property <schema-update-remove-property>`

Author: it means "really really required" ;-)

   * - :ref:`Change property type but keep the name <schema-update-change-property-type-but-keep-name>`
     - Breaking
     - Breaking

>    * - :ref:`Change property's status between from optional nd required <schema-update-change-property-status>`

Author: Overall, I was (am) aiming for simplification of the page. It just felt so overwhelming, especially since so much of the content was duplicated. This seems like a good place to reduce the duplication.

>    * - :ref:`Change property's status between from optional nd required <schema-update-change-property-status>`

Reviewer: Alright, that's fair. There's still a word choice issue, though - we should pick either "between" or "from". (I can't seem to fix that on this thread since it's outdated.)

adding the new property to each impacted document and populating it with a 
default value. After you've updated the documents, they will sync to the client 
application.

When you add a new required property, App Services retrieves the documents for 
the collection that have new values per the updated schema. App Services iterates 
through those documents and re-inserts them to get the new values. This is
expected behavior, but it does cause a temporary halt to propagating changes 
while this process is underway.

.. important::

   App Services uses a ``__realm_sync.unsynced_documents`` collection to track 
   unsynced documents. When you add a required property, the re-sync process can 
   push this collection over the limit of 100,000 documents. In this case, you must 
   :ref:`terminate and re-enable sync <terminating-realm-sync>`, even though the 
   change you're making is a non-breaking change. 

.. _schema-update-add-optional-property:

Add an Optional Property
------------------------

If you want to add an optional property to both the server-side schema and 
the client object model, you can add the optional property to the object 

> model, and use :ref:`Development Mode <development-mode>` to let 

Author: I'm sorry; I'm not sure what you're proposing here.

> model, and use :ref:`Development Mode <development-mode>` to let 

Reviewer: I'm proposing we do here below in ln 193 (which I can't comment on directly as we're not modifying it in this PR) what we did above in what is now ln 174 - consolidate the two callouts and make it an "Important" callout about this possibly requiring terminating & re-enabling Sync even though this is a non-breaking change. i.e. change the note below to:

> .. important::
> 
>    App Services uses a ``__realm_sync.unsynced_documents`` collection to track 
>    unsynced documents. When you add an optional property, the re-sync process can 
>    push this collection over the limit of 100,000 documents. In this case, you must 
>    :ref:`terminate and re-enable sync <terminating-realm-sync>`, even though the 
>    change you're making is a non-breaking change. 

(Assuming that's correct? I'm not 100% sure whether this applies - maybe only if you provide a default value for the optional?)

new objects retain the removed property, and App Services sets the value to an 
appropriate empty value, such as ``null`` for nullable properties, a 0 for 
integer values, or an empty string for string values. 

.. _schema-update-change-object-type-name:

Change an Object's Name
------------------------

Changing an object's name is a breaking change to both the server-side 
schema and the client-side object model.

Instead of changing an object's name, consider implementing the 
:ref:`partner collection strategy <schema-partner_collection-strategy>`, in which 
you keep the existing collection and schema, and create a new collection 
with the new schema. 

If you choose to change the object's name instead of using the partner collection 
strategy, you will need to terminate sync, manually update the schema, 
and re-enable sync. In addition, your client 
application must perform a :ref:`manual client reset <manual-client-reset>`.

.. note::

   :ref:`Development Mode <development-mode>` 
   does not automatically update your schema for breaking changes. 

.. _schema-update-change-property-name:

Change a Property Name
----------------------

> Changing a property name may be a breaking change to both the server-side 

Reviewer: If you map it, you're not changing the property name.

   to the client. This may give client users the impression that the data 
   has been lost.

Instead of changing an property's name, consider implementing the 
:ref:`partner collection strategy <schema-partner_collection-strategy>`, in which 
you keep the existing collection and schema, and create a new collection 
with the new schema. 

If you choose to change the property's name instead of using the partner collection 
strategy, you will need to terminate sync, manually update the schema, 
and re-enable sync. In addition, your client 
application must perform a :ref:`manual client reset <manual-client-reset>`.

When you terminate and re-enable sync, you must also update existing Atlas 
documents to enable them to Sync with your client applications. 
Without this additional handling, those documents do not Sync and it may 
appear in the client that the data has been lost. You could resolve this issue 
in two ways:

- Change the old field name on each document to match the new schema
- Add a new field to each document with a name that matches the new schema, 
  and copy the value from the old field into it

> After you've made these changes, the appropriate documents will 
> sync to the client application.

Reviewer: Same note as above - avoid tense changes and consider explaining why they sync again?

```suggestion
After you update the documents, they match the client schema and can sync again. 
```

   to the client. This may give client users the impression that the data 
   has been lost.

> Instead of changing an property's type, consider implementing the 

Reviewer: ```suggestion
Instead of changing a property's type, consider implementing the 
```

:ref:`partner collection strategy <schema-partner_collection-strategy>`, in which 
you keep the existing collection and schema, and create a new collection 
with the new schema. 

If you choose to change the property's type instead of using the partner collection 

> strategy, you will need to terminate sync, manually update the schema, 

Reviewer: ```suggestion
strategy, you must terminate sync, manually update the schema, 
```

- Add a new field to each document with the type that matches the new schema, 
  and copy the value from the old field into it, transforming its type

After you've made these changes, the appropriate documents should once again 
sync to the client application.

.. _schema-update-change-property-status:

Change a Property's Status Between Optional and Required
--------------------------------------------------------

Changing a property's status between optional and required 
is a breaking change to both the server-side schema and the client-side object model.

.. warning:: Update existing documents

   If you change a property's status in the server-side schema, you must 
   update existing documents with that new property type or they do not sync 
   to the client. This may give client users the impression that the data 
   has been lost.

Instead of changing an property's status, consider implementing the 
:ref:`partner collection strategy <schema-partner_collection-strategy>`, in which 
you keep the existing collection and schema, and create a new collection 
with the new schema. 

If you choose to change the property's status instead of using the partner collection 

> strategy, you will need to terminate sync, manually update the schema, 

Reviewer: ```suggestion
strategy, you must terminate sync, manually update the schema, 
```
